### PR TITLE
feat(`anvil`): migrate in-house `Genesis` to `alloy-genesis`

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

We were using a duped genesis struct for anvil.

## Solution

Switches anvil to use alloy-genesis

### CODE REVIEW

File: crates/anvil/src/config.rs [hunk 1]
```diff
@@ -834,7 +834,13 @@ impl NodeConfig {
...
         // if provided use all settings of `genesis.json`
         if let Some(ref genesis) = self.genesis {
-            genesis.apply(&mut env);
+            env.cfg.chain_id = genesis.config.chain_id;
+            env.block.timestamp = U256::from(genesis.timestamp);
+            if let Some(base_fee) = genesis.base_fee_per_gas {
+                env.block.basefee = U256::from(base_fee);
+            }
+            env.block.number = U256::from(genesis.nonce);
+            env.block.coinbase = genesis.coinbase;
```
Comment (Evalir) [inline crates/anvil/src/config.rs:843]: this applied these fields conditionally before (see `genesis.apply()`)—but they're not conditional on alloy-genesis. Should check for any default values or just apply them like this?

**Review by mattsse — CHANGES_REQUESTED** [review]
lets add private key to genesis account, geth also has these:

https://github.com/ethereum/go-ethereum/blob/fc380f52ef9778e988266f776b9593ce719cf79d/core/genesis.go#L200-L200

File: crates/anvil/core/src/eth/transaction/optimism.rs [hunk 2]
```diff
@@ -229,6 +229,14 @@ impl Transaction for DepositTransactionRequest {
...
         self.value
     }
+
+    fn encode_for_signing(&self, out: &mut dyn alloy_rlp::BufMut) {
+        self.encode_for_signing(out)
+    }
+
+    fn payload_len_for_signature(&self) -> usize {
+        self.payload_len_for_signature()
+    }
```
Comment (Evalir) [inline crates/anvil/core/src/eth/transaction/optimism.rs:239]: unrelated, comes from the alloy bump, see https://github.com/alloy-rs/alloy/pull/161

File: crates/anvil/src/config.rs [hunk 3]
```diff
@@ -834,7 +834,13 @@ impl NodeConfig {
 
         // if provided use all settings of `genesis.json`
         if let Some(ref genesis) = self.genesis {
-            genesis.apply(&mut env);
+            env.cfg.chain_id = genesis.config.chain_id;
+            env.block.timestamp = U256::from(genesis.timestamp);
+            if let Some(base_fee) = genesis.base_fee_per_gas {
+                env.block.basefee = U256::from(base_fee);
+            }
+            env.block.number = U256::from(genesis.nonce);
```
Comment (mattsse) [inline crates/anvil/src/config.rs:842]: this is not quite right

I think we're also missing optional number on alloy genesis:

https://github.com/ethereum/go-ethereum/blob/fc380f52ef9778e988266f776b9593ce719cf79d/core/genesis.go#L64-L64

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/anvil/src/config.rs (hunk 3 above)
Comment (Evalir) [reply crates/anvil/src/config.rs:842]: yep, let's add it.

Comment (Evalir) [reply crates/anvil/src/config.rs]: https://github.com/foundry-rs/foundry/pull/6970/commits/f0df51e06598db282cd630b8ab050015fdfb06de

**Comment by Evalir** [comment]
ci failure unrelated

**Review by mattsse — APPROVED** [review]

**Pull request merged**